File: packages/hht_auth/hht_auth_server/lib/src/repositories/sponsor_pattern_repository.dart
/// IMPLEMENTS REQUIREMENTS:
///   REQ-d00079: Linking Code Pattern Matching - Repository interface
///
/// Abstract repository interface for sponsor pattern data access.

import 'package:hht_auth_core/hht_auth_core.dart';

/// Repository interface for SponsorPattern persistence operations.
abstract class SponsorPatternRepository {
  /// Retrieves all active sponsor patterns.
  ///
  /// Results are cached for 5 minutes to improve performance.
  /// Patterns are sorted by prefix length descending (most specific first).
  Future<List<SponsorPattern>> getAllActivePatterns();

  /// Finds a sponsor pattern by linking code prefix match.
  ///
  /// Returns null if no matching pattern found or pattern is inactive.
  Future<SponsorPattern?> findByLinkingCode(String linkingCode);

  /// Creates a new sponsor pattern.
  Future<void> createPattern(SponsorPattern pattern);

  /// Decommissions a sponsor pattern (sets active=false).
  Future<void> decommissionPattern(String sponsorId);

  /// Refreshes the pattern cache.
  Future<void> refreshCache();
}

File: packages/hht_auth/hht_auth_server_ts/src/repositories/sponsor-pattern.repository.ts
/**
 * IMPLEMENTS REQUIREMENTS:
 *   REQ-d00079: Linking Code Pattern Matching
 *
 * Sponsor pattern repository interface and in-memory implementation.
 */

import {
  SponsorPattern,
  findMatchingPattern,
} from '../models/sponsor-pattern.model.js';

/**
 * Repository interface for sponsor pattern operations.
 */
export interface SponsorPatternRepository {
  getAllActivePatterns(): Promise<SponsorPattern[]>;
  findByLinkingCode(linkingCode: string): Promise<SponsorPattern | null>;
  findBySponsorId(sponsorId: string): Promise<SponsorPattern | null>;
}

/**
 * In-memory sponsor pattern repository for testing.
 */
export class InMemorySponsorPatternRepository
  implements SponsorPatternRepository
{
  private patterns: SponsorPattern[] = [];

  async getAllActivePatterns(): Promise<SponsorPattern[]> {
    return this.patterns.filter((p) => p.active);
  }

  async findByLinkingCode(linkingCode: string): Promise<SponsorPattern | null> {
    return findMatchingPattern(linkingCode, this.patterns);
  }

  async findBySponsorId(sponsorId: string): Promise<SponsorPattern | null> {
    return (
      this.patterns.find(
        (p) => p.sponsorId === sponsorId && p.active
      ) ?? null
    );
  }

  /** Clear all patterns (for testing) */
  clear(): void {
    this.patterns = [];
  }

  /** Seed patterns (for testing) */
  seed(patterns: SponsorPattern[]): void {
    this.patterns = [...patterns];
  }
}
